### ISSUE TITLE
Support for pre-declared "external" globals for sourced scripts to avoid SC2154 warnings.

### ISSUE THREAD

**Issue by deryni** [koalaman/shellcheck#356]
We have a series of scripts that operate on a set of shared variables that are created by the top-level script.

The secondary scripts use these variables without declaring them as they are known to exist when run.

shellcheck, obviously, doesn't know this. It would be nice if we could tell it about acceptable external global variables to disable SC2154 warnings about them.

I realize we could use `# shellcheck disable=SC2154` for them but that's a larger hammer than I would like as it will also disable other warnings from typos and the like if they happen to appear on that line.

The `context=sourced` suggestion from https://github.com/koalaman/shellcheck/issues/281 is somewhat related to this request.

**Comment by koalaman** [comment]
Global variables are conventionally in all uppercase (`PATH, DISPLAY, SSH_CLIENT`, etc). For this reason, ShellCheck generally doesn't warn when using undefined uppercase variables. 

The best solution would probably be for ShellCheck to process sourced files though.

**Comment by deryni** [comment]
That's true for system variables but I'm not sure that's as true for script-internal global variables (but maybe it is though that feels like playing in the system's/shell's playground a little bit).

Processing sourced files would be interesting though, assuming you mean when you pass shellcheck the toplevel file it would "recurse" into the sourced files and consider them in that context.

That wouldn't solve all of my cases (we have some scripts that get sourced and others that get run that all expect the external globals) but it would solve some of them.

**Comment by Moeriki** [comment]
When writing scripts for preset systems this uppercase convention can't always be adhered to.

Case in point. I'm writing scripts for Atlassian Bamboo. There is a global variable `bamboo` you can refer to for certain configurational values.

**Comment by deryni** [comment]
Bamboo uses many variables with the `bamboo_` prefix but not a single variable. (Though using a single associative array variable would certainly be nice.) But yes, same idea.

**Comment by koalaman** [comment]
I ran into the same issue with a library set `FLAGS_verbosity` style variables. 

How about a flag for setting a pattern/regex to ignore? This would let you do something like

    export SHELLCHECK_OPTS="--ignore-variable-pattern=FLAGS_|bamboo"

to not get these warnings anymore. Alternatively as a directive, starting the file with

    # shellcheck ignore-variable-pattern=FLAGS_|bamboo

**Comment by deryni** [comment]
Seems reasonable and would work for me. Though I'd probably suggest making it require explicit pattern markings (or having both pattern and non-pattern versions so that literal variables could be included in the list safely as well (i.e.

    --ignore-unassigned-variables='FLAGS_*|bamboo_*|explicit_global'
or

    --ignore-unassigned-variable-patterns='FLAGS_|bamboo_' --ignore-unassigned-variables=explicit_global

That said it also might make sense to handle this more generically (the way we discussed in #380, that is with names/patterns attached to the normal directives).

As the link in my comment on #380 is apparently now dead here's an [updated link](http://luacheck.readthedocs.org/en/0.12.0/cli.html#patterns) to the luacheck documentation for how it handles this idea.

The gist is that in addition to a directive like `# shellcheck disable=SC2154` you could use `# shellcheck disable=SC2154/FLAGS_*` or `# shellcheck disable=SC2154/bamboo_*` or `# shellcheck disable=SC2154/explicit_global` to scope the disabling to that specific variable/pattern. (I realize that is likely a fair bit more complex of an addition though. Though if my haskell was anything but non-existent I'd be interested in taking a crack at it.)

**Comment by Moeriki** [comment]
> Bamboo uses many variables with the bamboo_ prefix but not a single variable.

@deryni That depends. When running inline scripts it's a single variable. ${bamboo.myvar}
When running linked scripts it's all separate. ${bamboo_myvar}

> Bamboo variables are exported as bash shell variables. All full stops (periods) are converted to underscores. For example, the variable bamboo.my.variable is $bamboo_my_variable in bash. This is related to File Script tasks (not Inline Script tasks).
- https://confluence.atlassian.com/bamboo/bamboo-variables-289277087.html

**Comment by deryni** [comment]
@Moeriki `${bamboo.myvar}` isn't a shell variable. It isn't a legal variable name. (And even if it was that would still be many variables and not a single variable as that isn't how you index into a shell array variable that would be `${bamboo["myvar"]}`.)

What's happening there is that bamboo is pre-processing the script for you and replacing that (and I didn't know that worked in inline script bodies, I've never used it, but did know it worked in other places in bamboo itself).

The fact that it works in inline script bodies is sort of scary actually (likely safe-ish and possibly even entirely safe but still not something I want). Just one more reason for me to work on removing them from my builds entirely. =)

**Comment by Moeriki** [comment]
@deryni Ah I see. Thanks for the insight!

@koalaman your proposition would work fine for me.

**Comment by dflock** [comment]
Maybe it would be possible to use a new directive to point to the parent script, where the variables are defined?

**Comment by tomprince** [comment]
For example, for [this file](https://hg.mozilla.org/comm-central/file/36b19d54366f/comm-confvars.sh), I get (slightly reformated):

```
   7:18  warning  srcdir is referenced but not assigned.                    2154 (shellcheck)
  13:19  warning  _objdir is referenced but not assigned.                   2154 (shellcheck)
  18:5   warning  moztopsrcdir appears unused. Verify it or export it.      2034 (shellcheck)
  19:5   warning  commtopsrcdir appears unused. Verify it or export it.     2034 (shellcheck)
  21:5   warning  mozreltopsrcdir appears unused. Verify it or export it.   2034 (shellcheck)
  22:5   warning  commreltopsrcdir appears unused. Verify it or export it.  2034 (shellcheck)
  24:5   warning  commtopobjdir appears unused. Verify it or export it.     2034 (shellcheck)
```

I would like to specify something like
```
# shellcheck external_vars="srcdir _objdir"
# shellcheck exported_vars="moztopsrcdir commtopsrcdir mozreltopsrcdir commreltopsrcdir commtopobjdir"
```
to be explicit about what variables are defined by the sourcing script, and which are used by the sourcing script.

**Comment by mat813** [comment]
In the FreeBSD ports tree, we use environment variables extensively to pass values from the make(1) framework to the shell scripts doing the more complex stuffs, and I am in the process of [cleaning things up](https://reviews.freebsd.org/D13970) it would be nice to have a directive where one could add an exhaustive list of already existing variables, to be able to turn SC2154 on.

**Comment by matthewpersico** [comment]
Am I missing something? Seems to me that if you allow 

    # shellcheck external_vars="srcdir _objdir"

in one file, you're gonna need it in others. And then you'll get the bright idea to put the common declarations in one file, which you'll need to...

wait for it...

source in! :-)

So instead of a directive, how about just dropping a bunch of declare statements in a file foo.bashinc and sourcing that in with a hard coded path?

**Comment by mat813** [comment]
So, your idea is nice, for people who have variables that are the same in all their scripts.

It is absolutely not the case for the FreeBSD ports framework usage, a directive is really the only way to go for us to be able to use SC2154.

**Comment by Screwtapello** [comment]
Another use-case: The editor [Kakoune](https://kakoune.org/) uses POSIX sh for scripting, and [exports various bits of internal state as shell variables](https://github.com/mawww/kakoune/blob/master/doc/pages/expansions.asciidoc#shell-expansions). All these variables follow the pattern `kak_*` so it would be really handy to be able to tell ShellCheck to ignore variables matching that pattern.

**Comment by jasonkarns** [comment]
Another use case: Bats, which sets `status`, `output` and `lines` (https://github.com/sstephenson/bats#run-test-other-commands)

(related to #417 as a directive to silence bats-specific variables would aid in shellcheck supporting bats files.)

**Comment by dermoth** [comment]
Hi... I have a similar requirement, however since the script is closely tied to another I was rather thinking to tell shellcheck to "source" the file... I tried adding a `source=<otherfile>` but it's a no-op unless followed by something like `. /dev/null`

Wouldn't it make sense to allow `source=...` just as a hint of pre-declared variables, without actually sourcing it. Or maybe even better cross-link source file and parent and scope it from there, ex (maybe with something else than source in the 2nd script...):

#### `main.sh`
```bash
#!/bin/bash

myvar=foo

for f in inc/*.sh
do
    #shellcheck source=inc/foo.sh source=inc/bar.sh source=inc/baz.sh#  BTW wildcards would be nice here....
    . "$f"
done

myprivatevar=bar

```

#### `foo.sh`
```bash
# because we don't actually source main.sh but main.sh sources us, scope this script from matching main.sh source
# shellcheck shell=bash source=../main.sh # (or maybe something else like parent=../main.sh)

othervar=$myvar       # OK - already defined before main.sh source
echo "$myprivatevar"  # ERROR - not in scope
```

**Comment by abouvier** [comment]
Could it also be possible to support `source=path/to/script.sh` directive in a `.shellcheckrc` file, with `script.sh` containing all the exported variables.

**Comment by Jonta** [comment]
> Could it also be possible to support `source=path/to/script.sh` directive in a `.shellcheckrc` file, with `script.sh` containing all the exported variables.

Hopefully not in home. Ref. https://wiki.archlinux.org/index.php/XDG_Base_Directory =)

**Comment by mnuccioarpae** [comment]
> Hi... I have a similar requirement, however since the script is closely tied to another I was rather thinking to tell shellcheck to "source" the file... I tried adding a `source=<otherfile>` but it's a no-op unless followed by something like `. /dev/null`

@dermoth I have tried this approach, but it does not work because it creates an infinite loop.

For example, if Shellcheck starts from the main script, then sources the child script and then sources the main script again. The same thing happens if you start from the child script. I imagine Shellcheck can detect the loop, but the problem is that you are passing to Shellcheck the wrong information: the child script never sources the main script during execution.

**Comment by dermoth** [comment]
Can you break when you have already read a file?

Or maybe a "sourced by" directive, where you start form the sourced file 
up to the line or directive that sources the current script.

**Comment by mnuccioarpae** [comment]
I think a "sourced by" directive is cleaner because it expresses what happens during script execution.

But what do you do if the sourcing script is unknown or when is not a script that defines the external variables?

For now, I am using the workarounds suggested by @koalaman at https://github.com/koalaman/shellcheck/issues/2356#issuecomment-944608623

**Comment by mnuccioarpae** [comment]
I forgot to mention a workaround for bash that I find very useful because it's succinct while making the script fail fast when the needed environment is not fully defined.

I use "${external_var?}" instead of "$external_var" the **first** time i use "external_var" in the script. This way, you avoid the SC2154 warning, and the script fails if "external_var" is unexpectedly undefined.

Do you think I can add all these workarounds to https://www.shellcheck.net/wiki/SC2154?
It was not easy for me to find them.

**Comment by ShamrockLee** [comment]
Another use case: [Nixpkgs](https://github.com/NixOS/nixpkgs) [setup hooks](https://nixos.org/manual/nixpkgs/stable/#ssec-setup-hooks).

Nixpkgs is a package repository of Nix package manager, and setup hooks are shell scripts meant to be sourced when building a package. There are dependency relation between setup hooks, and all hooks depend on a base `setup.sh`, assuming the latter has been sourced in advance.

It would be great to have a way to hint ShellCheck that some scripts has been sourced before this one.

**Comment by ShamrockLee** [comment]
Or, as a workaround, this would also be achievable by supporting a per-file `check-sourced` directive. This way, one can write a meta-script, sourcing the external scripts without checking, and eventually source the target script and specify that it be checked.

**Comment by ingenarel** [comment]
i was configuring shellscript for ebuild files and approached a different way of fixing SC2034 and SC2154 because i misread something, however i do think implementing the feature could be a good idea, because it would fix this issue, https://github.com/koalaman/shellcheck/pull/2704, and probably a few other headaches

i thought using `source=<filename>` in the .shellcheckrc would make shellcheck source a file beforehand, so i tried to use that and eventually reread the wiki and realized that it was to dictate where to find the to-be-sourced file

i think implementing a cli arg or an option in the rcfile that just sources header like files could be a good idea, i don't know haskell otherwise i would've tried to implement it myself
